crates/synthesis/src/qsd.rs

```diff
@@ -0,0 +1,800 @@
+fn _block_zxz_decomp(
+    mat: &DMatrix<Complex64>,
+) -> (
+    DMatrix<Complex64>,
+    DMatrix<Complex64>,
+    DMatrix<Complex64>,
+    DMatrix<Complex64>,
+) {
+    debug_assert!(verify_unitary(mat));
+
+    let i = Complex64::new(0.0, 1.0);
+    let n = mat.shape().0 / 2;
+    let x = mat.view((0, 0), (n, n));
+    let y = mat.view((0, n), (n, n));
+    let u21 = mat.view((n, 0), (n, n));
+    let u22 = mat.view((n, n), (n, n));
+    let (sx, ux) = _zxz_decomp_svd(x.into_owned());
+    let (sy, uy) = _zxz_decomp_svd(y.into_owned());
+    let c = ((&uy.adjoint() * &ux) * i).adjoint();
+    let a1 = (sx + sy * i) * &ux;
+    let a2 = u21 + (u22 * (uy.adjoint() * ux) * i);
+    let b = (a1.adjoint() * x) * Complex64::from(2.0) - DMatrix::<Complex64>::identity(n, n);
+    (a1, a2, b, c)
+}
+
+/// Verify ZXZ decomposition gives the same unitary
+#[allow(clippy::toplevel_ref_arg)]
+fn _zxz_decomp_verify(
+    mat: &DMatrix<Complex64>,
+    a1: &DMatrix<Complex64>,
+    a2: &DMatrix<Complex64>,
+    b: &DMatrix<Complex64>,
+    c: &DMatrix<Complex64>,
+) -> bool {
+    let n = mat.shape().0 / 2;
+    let iden = DMatrix::<Complex64>::identity(n, n);
+
+    let a_block = stack![a1, 0; 0, a2];
+
+    let b1 = iden.clone() + b;
+    let b2 = iden.clone() - b;
+    let b_block = stack![b1, b2; b2, b1];
+
+    let c_block = stack![iden, 0; 0, c];
+
+    let mat_check = a_block * b_block * c_block * Complex64::from(0.5);
+
+    const EPS: f64 = 1e-7;
+    abs_diff_eq!(mat, &mat_check, epsilon = EPS)
+}
+
+///  Decompose a generic multiplexer.
+///
+///        ────□────
+///         ┌──┴──┐
+///       /─┤     ├─
+///         └─────┘
+///
+/// represented by the block diagonal matrix
+///
+///         ┏         ┓
+///         ┃ um0     ┃
+///         ┃     um1 ┃
+///         ┗         ┛
+///
+/// to
+///            ┌───┐
+///     ───────┤ Rz├──────
+///       ┌───┐└─┬─┘┌───┐
+///     /─┤ w ├──□──┤ v ├─
+///       └───┘     └───┘
+///
+/// where v and w are general unitaries determined from decomposition.
+///
+/// .. note::
+///
+///     When the original unitary is controlled then the default value of ``opt_a2 = False``
+///     as we start with the demultiplexing step that does not work with the optimization A.2 of [1, 2].
+fn demultiplex(
+    um0: &DMatrix<Complex64>,
+    um1: &DMatrix<Complex64>,
+    opt_a1: bool,
+    opt_a2: bool,
+    depth: usize,
+    _ctrl_index: Option<usize>,
+    vw_type: VWType,
+) -> PyResult<(CircuitData, DMatrix<Complex64>, DMatrix<Complex64>)> {
+    let um0 = closest_unitary(um0.clone());
+    let um1 = closest_unitary(um1.clone());
+
+    let dim = um0.shape().0 + um1.shape().0;
+    let num_qubits = dim.ilog2() as usize;
+    let _ctrl_index = _ctrl_index.unwrap_or_else(|| num_qubits - 1);
+    let layout: Vec<Qubit> = (0.._ctrl_index)
+        .chain(_ctrl_index + 1..num_qubits)
+        .chain([_ctrl_index])
+        .map(Qubit::new)
+        .collect();
+    let um0um1 = um0.clone() * um1.adjoint();
```

Comment:
You can avoid cloning by writing (not sure how to do this in 1 line):
```
let um0_view: DMatrixView<Complex64> = um0.as_view();
let um0um1 = um0_view * um1.adjoint();
```

Reply:
We can do it as a one-liner by doing
```
let um0um1 = &um0 * um1.adjoint();
```

The `adjoint()` still creates a copy; we may be able to optimize this away as well, but I don't know how - nalgebra supports [ad_mul](https://docs.rs/nalgebra/latest/nalgebra/base/struct.Matrix.html#method.ad_mul) which does `self.adjoint() * rhs`, which is the opposite of what we need. Really strange there's nothing for the other case.